tools/rostopic/src/rostopic/__init__.py

```diff
@@ -173,40 +176,123 @@ def print_hz(self):
             max_delta = max(self.times)
             min_delta = min(self.times)
 
-            self.last_printed_tn = self.msg_tn
-        print("average rate: %.3f\n\tmin: %.3fs max: %.3fs std dev: %.5fs window: %s"%(rate, min_delta, max_delta, std_dev, n+1))
+            self.last_msg_tn = self.msg_tn
+
+        return rate, min_delta, max_delta, std_dev, n+1
+
+    def print_hz(self):
+        """
+        print the average publishing rate to screen
+        """
+        if not self.times:
+            return
+        ret = self.get_hz()
+        if ret is None:
+            print("no new messages")
+            return
+        rate, min_delta, max_delta, std_dev, window = ret
+        print("average rate: %.3f\n\tmin: %.3fs max: %.3fs std dev: %.5fs window: %s"%(rate, min_delta, max_delta, std_dev, window))
 
 def _sleep(duration):
     rospy.rostime.wallsleep(duration)
 
-def _rostopic_hz(topic, window_size=-1, filter_expr=None):
+
+def get_parent_topics(topic, parent_topics=None, impl=False):
+    """Get topics to which the specified topic subscribes to.
+
+    @type topic: ``str``
+    @param topic: topic name
+    @type parent_topics: ``NoneType`` or ``list``
+    @param parent_topics: list of parent topics which already got.
+    @type impl: ``bool``
+    @param impl: search parent topics recursively or not
+    """
+    master = rosgraph.Master('/rostopic')
+    publications, subscriptions, _ = master.getSystemState()
+    if parent_topics is None:
+        parent_topics = []
+    pub_nodes = [l for t, l in publications if t == topic]
+    if not pub_nodes:
+        return parent_topics
+    for node in pub_nodes[0]:
+        subs = [t for t, l in subscriptions if node in l]
+        subs = filter(lambda x: x not in parent_topics, subs)  # get unique
+        parent_topics.extend(subs)
+        if not impl:
+            return parent_topics
+        for sub in subs:
+            parent_topics = get_parent_topics(sub, parent_topics, impl=impl)
```

Comment:
Before calling this function recursively it should be checked that `sub` is not alread in `parent_topcis`. While that condition was true a few lines above after other recursive calls it might not anymore.